### PR TITLE
handle http auth

### PR DESCRIPTION
**Pull request by reepush**
Finally got it working for both Chrome and Firefox!

* In emulation mode callback for _mooltipass.device.retrieveCredentials_ didn't execute that's why there was no auth popup.
* Stuck with a problem that it's impossible to get `Authorization` header while we intercept response. But we can create our custom popup and handle http authorization by ourself. I used window.prompt now, will add popup later.
* Firefox doesn't execute callback because of the _asyncBlocking_ setting. It's not supported and I will see the differences between Chrome's _asyncBlocking_ and Firefox's _blocking_.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -14,45 +14,43 @@ env:
...
+  - secure: fVwYFsZkADhlLb742KRoxIxRw5MH2gAX3Pxa2YQvJLFh0pkkU/61GTVIhf3yReXAwaRcjES7NzPaqCguT7d19Ak7oD5Uw3dO9q/ZEGd7bDUNT2v+VLeECkY+R/ZKTWTyzAA+IQLR3SuqWAU8i8ocq6LNTPSHD5px0h2ofCapHpFUUuKt2lyEonxjdqRgbjexEha3UOHcC6uvQicFE8AfDGHvDgk1gtRQTYKKLG9+eMxNexVZvc5w5jSlWREVTxula/PqI1qmX/QUDmjmTTmS37cxr0AuAeDQwYoO3NPC+n3Cl0mgcQR33n1bdTMFTdkRPQgffSuCoIpF+tP/DI1mnDi8VgVRRbL90TEwPFw45lHzd//ChZ4eI1/3R+gkZ2hBzlf1IcPzHmJpC+0/MYjO9PjdlutdBJD4yfaiCbi4kaBNqL/UNkmgjwFd18UfLudLkfVZ1+Ug7v5Y/dgPnDd1I48ry9WLOSjn4WkDX1CdA7iC3w8gGHPVOZAvqs6HkzabiNa37xbJrEL2oAMQNxRiy0uKQForNsRJH9bkNIeHlTwKKUKiLq08/7O55MXMwLLB++Rk/oJLo0hY0V3BoDOP0pxt8iEmU17cSS0IzPty0jfq5lruC+2rIrdvf5JA22rKCyIN3lYZoGQia6CBTMCc91tLwOSgBMIJ6igyhQ0pex8=
+script: if [[ $TRAVIS_COMMIT_MESSAGE == *"[ci-ext]"* ]]; then mvn install -Dcucumber.options="src/test/resources/mooltipass/automatedTest/features/$TEST_SUITE"; fi ;
```
Comment (limpkin) [inline .travis.yml]: please don't push your travis.

File: chrome_extension/vendor/mooltipass/mcCombinations.js [hunk 2]
```diff
@@ -140,6 +140,15 @@ var extendedPost = {
 	}
 }
 
+var extendedSubmitButton = {
+	'vodafone.de': function(form) {
+		return form.parent().parent().find('#loginBtn')
+	},
+	'nordinary-com.netcup-mail.de': function(form) {
+		return form.find('#submit')
+	},
```
Comment (limpkin) [inline chrome_extension/vendor/mooltipass/mcCombinations.js]: didn't we agree that we'd try another approach than this one?

File: chrome_extension/vendor/mooltipass/mcCombinations.js [hunk 3]
```diff
@@ -941,9 +950,17 @@ mcCombinations.prototype.doSubmit = function doSubmit( currentForm ) {
...
+		
 		// Try to click the submit element
 		var submitButton = currentForm.element.find(':submit:visible');
+		
+		// Run special cases
+		for ( specialCase in extendedSubmitButton ) {
+			if ( window.location.hostname.indexOf( specialCase ) > -1 ) {
+				submitButton = extendedSubmitButton[specialCase](currentForm.element);
+			}
+		}
```
Comment (limpkin) [inline chrome_extension/vendor/mooltipass/mcCombinations.js]: same as above.

File: chrome_extension/background/httpauth.js [hunk 4]
```diff
@@ -69,6 +69,14 @@ httpAuth.loginOrShowCredentials = function (logins) {
     }
     // no logins found
     else {
-        httpAuth.callback({});
```
Comment (limpkin) [inline chrome_extension/background/httpauth.js]: does that mean that if we don't know a given credential, we'll get these prompts on chrome, firefox AND safari?
this is not ok.

File: chrome_extension/vendor/mooltipass/device.js [hunk 5]
```diff
@@ -329,6 +335,22 @@ mooltipass.device.sendCredentialRequestMessageFromQueue = function()
                     }
                 }
                 if (background_debug_msg > 3) mpDebug.log("%c Emulation mode: nothing in buffer!", mpDebug.css('00ff00'));
+                
+                if (!credentialsFound) {
```
Comment (limpkin) [inline chrome_extension/vendor/mooltipass/device.js]: wait... if we actually find the credentials, don't we return from the function? (line 328/334)
your !credentialsFound is therefore not necessary, right?

File: chrome_extension/vendor/mooltipass/mcCombinations.js (hunk 2 above)
Comment (reepush) [inline chrome_extension/vendor/mooltipass/mcCombinations.js]: I branched from master, that's why these changes are here. Don't mention them.

File: chrome_extension/background/httpauth.js (hunk 4 above)
Comment (reepush) [inline chrome_extension/background/httpauth.js]: We will show custom popup for user with login and password inputs.

> But we can create our custom popup and handle http authorization by ourself. I used window.prompt now, will add popup later.

Comment (limpkin) [inline chrome_extension/background/httpauth.js]: can't you restrict your fix for firefox only?

File: chrome_extension/vendor/mooltipass/device.js (hunk 5 above)
Comment (reepush) [inline chrome_extension/vendor/mooltipass/device.js]: That's right, didn't mention it. Thanks.

File: chrome_extension/background/httpauth.js (hunk 4 above)
Comment (reepush) [inline chrome_extension/background/httpauth.js]: We need custom popup for both Chrome and Firefox. Look at my comment above.

**Comment by reepush** [comment]
Force pushed to _http-auth_ branch. Please check the changes.

screen shot 2017-07-06 at 20 57 56

* Added custom overlay for http auth
* Handled Firefox spec
* Made changes to _credentialsFound_ as discussed previously

What we need to handle also:
* Pressing _Cancel_ in popup should stop request
* Show error after submitting invalid credentials
* Also remember new credentials after invalid ones were submitted
* Handle proxy authorization
* Handle multiple http authorization at the same time

Last two issues relates to the fact that I removed a bunch of code from _background/httpauth.js_. For my test case everything works and sufficient. But I think I need to put some of the code back to solve last issues. Anyway I need to understand what's needed and what's not and now it's easier to add code step by step.

File: chrome_extension/mooltipass-content.css [hunk 6]
```diff
@@ -904,4 +907,134 @@ input.mp-hover-password {
 
 .mp-genpw-dialog .login-area {
     display: none;
-}
\ No newline at end of file
+}
+
+/* Overlay */
+
+.mp-overlay {
```
Comment (limpkin) [inline chrome_extension/mooltipass-content.css]: could you rename mp-overlay to something more explicit like mp-http-auth-overlay ?

File: .travis.yml (hunk 1 above)
Comment (limpkin) [inline .travis.yml]: can you remove this travis from your PR?

File: chrome_extension/background/httpauth.js [hunk 7]
```diff
@@ -1,74 +1,25 @@
...
-    // but in background.js only tab.id is used. To get tabs we could use
-    // chrome.tabs.get(tabId, callback) <-- but what should callback be?
-
-    var url = (httpAuth.isProxy && httpAuth.proxyUrl) ? httpAuth.proxyUrl : httpAuth.url;
-    //console.log('httpAuth: isProxy', httpAuth.isProxy, ' proxyUrl:', httpAuth.proxyUrl, 'url:', httpAuth.url);
-
-    mooltipass.device.retrieveCredentials(httpAuth.loginOrShowCredentials, {"id": details.tabId}, url, url, true);
-}
-
-httpAuth.loginOrShowCredentials = function (logins) {
```
Comment (limpkin) [inline chrome_extension/background/httpauth.js]: why did you have to remove all that code?

File: chrome_extension/vendor/mooltipass/device.js [hunk 8]
```diff
@@ -10,7 +10,7 @@ mooltipass.device = mooltipass.device || {};
 /**
  * Mooltipass emulation vars
  */
-mooltipass.device.emulation_mode = false
+mooltipass.device.emulation_mode = true
```
Comment (limpkin) [inline chrome_extension/vendor/mooltipass/device.js]: please leave it to false!

File: chrome_extension/mooltipass-content.js [hunk 9]
```diff
@@ -1572,7 +1572,7 @@ cip.rememberCredentials = function(event, usernameField, usernameValue, password
 			});
 		}
 
-		var url = mpJQ(this)[0].action;
+		var url = mpJQ(event.target)[0].action;
```
Comment (limpkin) [inline chrome_extension/mooltipass-content.js]: could you explain that change?

Comment (reepush) [inline chrome_extension/mooltipass-content.js]: This line suppose to get action attribute of the form, but this evaluates to `cip` instance and not to DOM element. Changed it to `event.target` and now we are able to save injected proxy address. Take a look at this code in commit:

```js
// Set proxy address as action attribute of the form, so mcCombs will
// save credentials for the right url.
if (req.args[0].isProxy) {
  mpJQ('.mp-popup-http-auth__form').attr('action', req.args[0].proxyURL)
}
```

**Comment by limpkin** [comment]
it seems to work well... congrats!
still waiting to read CI integration tests though
